### PR TITLE
feat(plugin-meetings): enable transcription event

### PR DESCRIPTION
**Pull request by akoushke**
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-243877

---

https://confluence-eng-gpk2.cisco.com/conf/display/WBXT/Webex+Browser+SDK+Real-time+Transcription

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/README.md [hunk 1]
```diff
@@ -342,6 +342,20 @@ webex.meetings.on('meeting:added', (addedMeeting) => {
...
 
+#### Enable to receive transcription payload live
+
+In order to receive the payload, Webex assistant must be enabled in the meeting.
+
+```javascript
+...
+// join to receive transcription payload
+meeting.join({enableTranscription: true}).then((res) => {
+  meeting.on('meeting:started:transcription', (payload) => {
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/README.md:353]: We wanted to move this before the `join` call.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 2]
```diff
@@ -2842,6 +2854,176 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * Check if the meeting supports the Webex Assistant feature
+   * @returns {boolean}
+   * @throws TranscriptionNotSupportedError
+   */
+  isTranscriptionSupported() {
+    if (this.isMeetingAdhoc()) {
+      throw new TranscriptionNotSupportedError(
+        'meeting does not support Webex Assistant feature'
+      );
+    }
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2875]: Remove this `if` statement

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+   * trigger and invoke the callback when the web socket connection is closed
+   *
+   * @param {function} callback
+   * @returns {void}
+   */
+  onCloseSocket(callback) {
+    this.webSocket.onclose = (event) => {
+      callback(event);
+    };
+  }
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/transcription/index.js:57]:
```suggestion
  onCloseSocket(callback) {
    if (callback) {
      this.webSocket.onclose = callback;
    }
  }
```

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+   * trigger and invoke when a web socket connection error occurs
+   *
+   * @param {function} callback
+   * @returns {void}
+   */
+  onErrorSocket(callback) {
+    this.webSocket.onerror = (event) => {
+      callback(event);
+    };
+  }
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/transcription/index.js:69]:
```suggestion
  onErrorSocket(callback) {
    if (callback) {
      this.webSocket.onerror = callback;
    }
  }
```

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * Get current speaker from the given csis payload
+   *
+   * @private
+   * @param {object} csis
+   * @returns {object}
+   */
+  getSpeaker(csis) {
+    return Object.values(this.members.membersCollection.members)
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/transcription/index.js:79]: Use `Map` to cache results

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      });
+  }
+
+  /**
+   * retrieve and pass the transcription inside the callback to return the payload
+   *
+   * @param {function} callback
+   * @returns {void}
+   */
+  get(callback) {
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/transcription/index.js:101]:
```suggestion
  subscribe(callback) {
```

**Review by mwegman — CHANGES_REQUESTED** [review]

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+              this.memberCSIs[member.id] = csi;
+              selfIsSpeaking = true;
+              break;
+            }
+          }
+        }
+
+        return selfIsSpeaking;
+      });
+  }
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/transcription/index.js:103]:
```suggestion
  getSpeaker(csis) {
    for (const csi of csis) {
      const member = this.memberCSIs[csi];
      if (member) {
        return member;
      }
    }

    return Object.values(this.members.membersCollection.members)
      .find((member) => {
        const memberCSIs = member.participant.status.csis;
        let selfIsSpeaking = false;

        for (const csi of csis) {
          if (memberCSIs.includes(csi)) {
            this.memberCSIs[csi] = member;
            selfIsSpeaking = true;
            break;
          }
        }

        return selfIsSpeaking;
      });
  }
```

**Review by mwegman — CHANGES_REQUESTED** [review]

File: packages/node_modules/samples/browser-plugin-meetings/index.html [hunk 8]
```diff
@@ -182,6 +198,10 @@ <h2>Meeting Controls</h2>
             <button id="gc-dtmf-tones-send" type="button" onclick="sendDtmfTones()">meeting.sendDTMF()</button>
             <label id="gc-dtmf-status"></label>
           </div>
+          <div class="transcription">
+            <button id="gc-start-receiving-transcription" disabled type="button" onclick="startReceivingTranscription()">start receiving transcription (click me before joining)</button>
+            <textarea id="gc-transcription-content" disabled>Transcription Content: Webex Assistant must be enabled, check the console!</textarea>
```
Comment (lalli-flores) [inline packages/node_modules/samples/browser-plugin-meetings/index.html:203]: The idea of this being a text area was that the transcripts would show on the app itself. if you're differing to just showing it in the console, that's fine but then there's no need for this text area

File: packages/node_modules/samples/browser-plugin-meetings/index.html [hunk 9]
```diff
@@ -182,6 +198,10 @@ <h2>Meeting Controls</h2>
             <button id="gc-dtmf-tones-send" type="button" onclick="sendDtmfTones()">meeting.sendDTMF()</button>
             <label id="gc-dtmf-status"></label>
           </div>
+          <div class="transcription">
+            <button id="gc-start-receiving-transcription" disabled type="button" onclick="startReceivingTranscription()">start receiving transcription (click me before joining)</button>
```
Comment (lalli-flores) [inline packages/node_modules/samples/browser-plugin-meetings/index.html:202]: you need to enable a way to toggle out of the subscription

**Review by lalli-flores — CHANGES_REQUESTED** [review]

File: packages/node_modules/samples/browser-plugin-meetings/index.html (hunk 8 above)
Comment (akoushke) [reply packages/node_modules/samples/browser-plugin-meetings/index.html:203]: im not logging out anymore. just the textarea for now.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 10]
```diff
@@ -808,7 +819,7 @@ export default class Meeting extends StatelessWebexPlugin {
    * @memberof Meeting
    */
   setUpLocusInfoListeners() {
-    // meeting update listeners
+    // meeting update listeners©
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:822]: You can remove this extra character

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  constructor(webSocketUrl, members, sessionId, token) {
+    this.webSocketUrl = webSocketUrl;
+    this.members = members;
+    this.sessionID = sessionId;
+    this.token = token;
+    this.memberCSIs = {};
+
+    const params = '?outboundWireFormat=text&bufferStates=true&aliasHttpStatus=true';
+
+    this.webSocket = new WebSocket(`${this.webSocketUrl}${params}`); // eslint-disable-line no-undef
```
[the remark this answers is not in the document]
Comment (mwegman) [reply packages/node_modules/@webex/plugin-meetings/src/transcription/index.js:24]: Will this code be running in node? If it is, we need to do something different. This line only works if it is being run in the browser.

**Comment by akoushke** [comment]
@mwegman  not in node. Only on browsers.

File: packages/node_modules/@webex/plugin-meetings/src/transcription/index.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (akoushke) [reply packages/node_modules/@webex/plugin-meetings/src/transcription/index.js]: not in node. only n the browser. @mwegman

[the remark this answers is not in the document]
Comment (akoushke) [reply packages/node_modules/@webex/plugin-meetings/src/transcription/index.js]: @lalli-flores added it globally.

**Review by lalli-flores — APPROVED** [review]
Thank you for contributing this feature 🙏🏼

**Pull request merged**